src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -85,54 +73,104 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
     }
 
     // see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs
-    private class PropertyFetch
+#if NET6_0_OR_GREATER
+    [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+#endif
+    private abstract class PropertyFetch
     {
-        public static PropertyFetch Create(TypeInfo type, string propertyName)
+        public static PropertyFetch? Create(TypeInfo type, string propertyName)
         {
-            var property = type.DeclaredProperties.FirstOrDefault(p => string.Equals(p.Name, propertyName, StringComparison.OrdinalIgnoreCase));
-            if (property == null)
-            {
-                property = type.GetProperty(propertyName);
-            }
-
+            var property = type.DeclaredProperties.FirstOrDefault(p => string.Equals(p.Name, propertyName, StringComparison.OrdinalIgnoreCase)) ?? type.GetProperty(propertyName);
             return CreateFetcherForProperty(property);
 
-            static PropertyFetch CreateFetcherForProperty(PropertyInfo propertyInfo)
+            static PropertyFetch? CreateFetcherForProperty(PropertyInfo? propertyInfo)
             {
                 if (propertyInfo == null || !typeof(T).IsAssignableFrom(propertyInfo.PropertyType))
                 {
                     // returns null and wait for a valid payload to arrive.
                     return null;
                 }
 
-                var typedPropertyFetcher = typeof(TypedPropertyFetch<,>);
-                var instantiatedTypedPropertyFetcher = typedPropertyFetcher.MakeGenericType(
-                    typeof(T), propertyInfo.DeclaringType, propertyInfo.PropertyType);
-                return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
-            }
-        }
+                var declaringType = propertyInfo.DeclaringType;
+                if (declaringType!.IsValueType)
+                {
+                    throw new NotSupportedException(
+                        $"Type: {declaringType.FullName} is a value type. PropertyFetcher can only operate on reference payload types.");
+                }
 
-        public virtual bool TryFetch(object obj, out T value)
-        {
-            value = default;
-            return false;
+                if (declaringType == typeof(object))
```

Comment:
Update on this: 
The branch will never be executed. It's here for the AOT-compiler (net7.0) so the trimmer does not remove this code.
Will update the comment accordingly.

Reply:
@agocke - can you confirm this?

Reply:
In general, the bar is pretty high for me to backport AOT changes to .NET 7 since it was intentionally a limited release. But even more importantly, .NET 7 is going out of support this year, so we almost certainly won't be backporting anything.

Reply:
Just to close on this discussion - as is, the code is unreachable and only there to hint the compiler to keep the instantiation for ref types (canon).
I think in theory this can be changed to use any other ref type - so for example `if (declatingType == typeof(Type))` woudl work as well - I picked `Type` since it's a public ref type which has properties - so this is potentially reachable. But I think using object is more "self explanatory".
I agree with @agocke that backporting the fix to the AOT compiler to 7 would be a hard sell - it's likely not a simple backport (the compiler changed a lot between 7 and 8) and this code is the only place where we ran into this (this code in OTel actually made us fix the compiler in 8).